### PR TITLE
Adjust stencil tests to use hardcoded python source opposed to AST.

### PR DESCRIPTION
**Pull request by stuartarchibald**
This patch comprises largely "mechanical" changes, the aim is to drop reliance on the Python 3.7 AST ahead of being able to drop Python 3.7 support entirely. To do this, all the use of the AST and manipulations of it have been replaced with hardcoded equivalent pure python functions. This diff touches almost the entire test file and so fixes to the remaining few flake8 violations were rolled into the patch at the same time.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Build `macOS py37_np118` failed with the seemingly unrelated:

<details>
<summary>Details:</summary>

```
2022-10-17T15:18:35.9326080Z ##[section]Starting: Build
2022-10-17T15:18:35.9338730Z ==============================================================================
2022-10-17T15:18:35.9339360Z Task         : Command line
2022-10-17T15:18:35.9339900Z Description  : Run a command line script using Bash on Linux and macOS and cmd.exe on Windows
2022-10-17T15:18:35.9340420Z Version      : 2.201.1
2022-10-17T15:18:35.9340810Z Author       : Microsoft Corporation
2022-10-17T15:18:35.9341380Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/command-line
2022-10-17T15:18:35.9341970Z ==============================================================================
2022-10-17T15:18:36.2468700Z Generating script.
2022-10-17T15:18:36.3291260Z ========================== Starting Command Output ===========================
2022-10-17T15:18:36.3340490Z [command]/bin/bash --noprofile --norc /Users/runner/work/_temp/e7d83009-2883-4de6-8494-04b5c2efbfa4.sh
2022-10-17T15:18:38.1615450Z 
2022-10-17T15:18:38.1617470Z # Build numba extensions without silencing compile errors
2022-10-17T15:18:38.1619800Z if [[ "$(uname -s)" == *"Linux"* ]] && [[ "$(uname -p)" == *"86"* ]]; then
2022-10-17T15:18:38.1622470Z     EXTRA_BUILD_EXT_FLAGS="--werror --wall"
2022-10-17T15:18:38.1623570Z else
2022-10-17T15:18:38.1624740Z     EXTRA_BUILD_EXT_FLAGS=""
2022-10-17T15:18:38.1626190Z fi
2022-10-17T15:18:38.1628120Z uname -s
2022-10-17T15:18:38.1646110Z 
2022-10-17T15:18:38.1647320Z if [[ $(uname) == "Darwin" ]]; then
2022-10-17T15:18:38.1648730Z     # The following is suggested in https://docs.conda.io/projects/conda-build/en/latest/resources/compiler-tools.html?highlight=SDK#macos-sdk
2022-10-17T15:18:38.1650340Z     wget -q https://github.com/phracker/MacOSX-SDKs/releases/download/11.3/MacOSX10.10.sdk.tar.xz
2022-10-17T15:18:38.1651560Z     shasum -c ./buildscripts/incremental/MacOSX10.10.sdk.checksum
2022-10-17T15:18:38.1652600Z     tar -xf ./MacOSX10.10.sdk.tar.xz
2022-10-17T15:18:38.1653180Z     export SDKROOT=`pwd`/MacOSX10.10.sdk
2022-10-17T15:18:38.1653720Z fi
2022-10-17T15:18:38.1654830Z uname
2022-10-17T15:18:43.6257750Z ##[error]Bash exited with code '8'.
2022-10-17T15:18:43.6283580Z ##[section]Finishing: Build
```

</details>

File: numba/tests/test_stencils.py [hunk 1]
```diff
@@ -2095,50 +1789,95 @@ def kernel(a):
...
         """cval is func expr"""
         def kernel(a):
             return a[0, 1] + a[0, -1] + a[1, -1] + a[1, -1]
+        # ----------------------------------------------------------------------
+        # Autogenerated kernel
+
+        def __kernel(a, neighborhood):
+            self.check_stencil_arrays(a, neighborhood=neighborhood)
+            __retdtype = kernel(a)
+            __b0 = np.full(a.shape, -0.2750268284872752, dtype=type(__retdtype))
```
Comment (stuartarchibald) [inline numba/tests/test_stencils.py:1805]: replace const with expr

Comment (stuartarchibald) [reply numba/tests/test_stencils.py]: Done in 186d079

**Review by apmasell — APPROVED** [review]
I hope these kernels don't need to be regenerated anytime soon.

**Review by sklam — CHANGES_REQUESTED** [review]
https://github.com/numba/numba/pull/8518/files#diff-8248fddc77eeb2725200fe3cdd541f4ddf04c81b9c472c33c082d0286b691475R80

the skip for 3.7 can be removed

**Comment by stuartarchibald** [comment]
> https://github.com/numba/numba/pull/8518/files#diff-8248fddc77eeb2725200fe3cdd541f4ddf04c81b9c472c33c082d0286b691475R80
> 
> the skip for 3.7 can be removed

@sklam thanks, addressed in 90cfa49.

@apmasell thanks for the review!

**Review by sklam — APPROVED** [review]

**Pull request merged**